### ISSUE TITLE
[Ivy] NgClass adding incorrect classes on subsequent CD runs

### ISSUE THREAD

**Issue by pshurygin** [angular/angular#33035]
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/common or @angular/core

### Is this a regression?
Yes, it works correctly with VE

### Description

I have encountered a weird issue with NgClass directive when testing Ivy update. I have a template like this:

```
<div class="main-content"
        [ngClass]="{'main-content_disabled': isDisabled}">
</div>
```

"isDisabled" variable is defined in the component and always equals boolean false in this particular case(and i have verified this using devtools).
On the initial page load without user interaction the result html is correct: a div with main-content class only. But when CD is triggered somewhere down the component tree via user interaction, which caused some DOM modifications in child components, main-content_disabled class gets added to the div (though the value of isDisabled property on the component is still false).

I have debugged into this a little bit by putting breakpoint on attribute modifications for this element and have probably discovered a cause for this issue(please look at the attached screenshot with devtools):
setClass method is being called with className="main-content_disabled" and value="false". String false and not a boolean false. This causes the wrong branch to be executed, adding class instead of removing it.
The transformation from false to "false" happened up in the callstack in the applyStylingValue method here:
```
if (isStylingValueDefined(valueToApply)) {
    valueToApply = sanitizer ? sanitizer(prop, value, 2 /* SanitizeOnly */) : valueToApply;
}
```
Sanitizer here transforms false into "false", which is probably not what should have happened.

Please let me know, if i could provide any additional info for this case.

## 🔥 Exception or Error
![ngClassIssue1](https://user-images.githubusercontent.com/25816676/66389930-2d10f200-e9d2-11e9-8569-605d70ec80c7.PNG)

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.0-next.8
Node: 10.16.0
OS: win32 x64
Angular: 9.0.0-next.9
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router

Package                                    Version
--------------------------------------------------------------------
@angular-devkit/architect                  0.900.0-next.8
@angular-devkit/build-angular              0.900.0-next.8
@angular-devkit/build-optimizer            0.900.0-next.8
@angular-devkit/build-webpack              0.900.0-next.8
@angular-devkit/core                       9.0.0-next.8
@angular-devkit/schematics                 9.0.0-next.8
@angular/cdk                               9.0.0-next.0
@angular/cli                               9.0.0-next.8
@angular/material                          9.0.0-next.0
@ngtools/webpack                           9.0.0-next.8
@nguniversal/module-map-ngfactory-loader   9.0.0-next.3
@schematics/angular                        9.0.0-next.8
@schematics/update                         0.900.0-next.8
rxjs                                       6.5.3
typescript                                 3.5.3
webpack                                    4.41.0
</code></pre>

**Comment by kara** [comment]
@pshurygin I can't seem to reproduce the behavior with the code snippet you provided. Can you create a minimal repro in the form of a Github repo or even a Stackblitz (it doesn't support Ivy, but I can port it to a repo that does)?

**Comment by pshurygin** [comment]
@kara The app itself is rather complex. I have tried to narrow down the issue, but stumbled upon another error, which could be the root cause(or it could be not).

The app itself is a renderer for a website based on the json configuration. It consists of widget-lists, which contain widgets. Some of the widgets may contain widget-list of its own with nested widgets inside. So we have a nested recursive component structure of any depth like this:
```
<widget-list>
    <widget>
        <widgets-list>
            <widget>...</widget>
        </widgets-list>
    </widget>
    <widget>...</widget>
</widget-list>
```
This actually works both with VE and Ivy. It looks like this recursive structure has something to do with the error, as it only appears when mouseover event is triggered on the nested widget, triggering some actions. 

But when i removed some code(like routing, backend json data resolving and more) to get to the issue root cause, i got circular dependancy warning during buildtime (like widgets-list -> widget -> [container widget type] -> widgets-list). This is expected and should not cause problems, but during runtime the app suddenly breaks with the error:
`Uncaught ReferenceError: Cannot access 'WidgetsListComponent' before initialization
    at Module.WidgetsListComponent (widgets-list.component.ts:1)
    at Module../src/app/renderer/widgets/columns/columns.component.ts (columns.component.html:12)
    at __webpack_require__ (bootstrap:84)
    at Module../src/app/renderer/widget/widget.component.ts (stripes-list.component.ts:29)
    at __webpack_require__ (bootstrap:84)
    at Module../src/app/renderer/widgets-list/widgets-list.component.ts (widgets-list.component.ts:1)
    at __webpack_require__ (bootstrap:84)
    at Module../src/app/app.component.ts (app.component.ts:1)
    at __webpack_require__ (bootstrap:84)
    at Module../src/app/app.module.ts (app.module.ts:1)`
So i'm not sure how to proceed any further. I have even removed all of the circular dependencies in our own code(in component.ts files), but the issue persists, as the dependency is actually still there on the template level(and angular is probably generating these dependencies internally anyway).

**Comment by pshurygin** [comment]
I will probably try to make a separate repo with this new error later today.

**Comment by pshurygin** [comment]
@kara Ok, i have managed to make a repo for an original issue. Turns out it is not related to circular dependencies, but rather a weird issue with multiple HostBindings + style sanitizer on a directive.
Here it is: 
https://github.com/pshurygin/ivy-ngclass-issue

**Comment by pshurygin** [comment]
@kara @JoostK Can you please take a look at the repo and remove the "needs repro" label?
It would be great if this one could be fixed before v9 release, as this is the only left issue preventing us from using ivy as far as i can tell. And it breaks our app in a bad way, adding incorrect classes all over the DOM.

**Comment by kara** [comment]
@pshurygin Thank you! I can reproduce the problem with the given repo. We'll put it on our roadmap.

**Comment by kara** [comment]
@pshurygin This should be resolved by 1cda80eb3a823af5c8fe1d9d9de55ad56eee770e, which should be in `next.11`
